### ISSUE TITLE
Load test/stress test for jobdb and possibly LDAP

### ISSUE THREAD

**Issue by mbenkmann** [mbenkmann/limux-gosa#8]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on September 10, 2012 21:09:13_

Write an external program that sends via parallel connections a mix of foreign_job_updates and gosa_query_jobdb. The test should fulfill 2 purposes:
- Determine how many requests per second go-susi can process
- Verify proper locking of the jobdb

The test should be able to run against an ordinary gosa-si, too.

_Original issue: http://code.google.com/p/go-susi/issues/detail?id=8_

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on September 14, 2012 14:22:45_

The test should also monitor memory consumption, so that it can be run over an extended period of time to see if there are memory leaks.

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on September 15, 2012 01:01:52_

Use runtime.NumGoroutine to spot goroutine leaks during the test. Maybe add a remote command that queries this data that would be called by 
go-susi --stats 
  similar to 
go-susi --sniff

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on September 15, 2012 01:05:35_

runtime.ReadMemStats would also be useful as part of go-susi --stats

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on November 13, 2012 09:37:22_

**Status:** Later

**Comment by mbenkmann** [comment]
_From [mux2...@gmail.com](https://code.google.com/u/114145867202573191452/) on November 14, 2012 04:25:57_

If LDAP access is rewritten to directly use libldap bindings a similar kind of load/stress test should be done. The following is a possible general algorithm for conducting either test:

k goroutines in parallel execute the following

open ldap connection
for {
  n = random uint
  for 1 to n {
    test_read = random bool
    na = random uint
    attr_list = list of na randomly chosen LDAP attributes
    no = random uint
    objlist = list of no randomly chosen LDAP objects
    if test_read {
      read attr_list attributes for all objlist objects
      write results into readlog with current timestamp
    } else {
      generate no*na random values
      write values to write log with current timestamp
      write new attribute values for all attr_list attributes for all objlist objects
    }
  }
  close ldap connection
  reopen ldap connection
}

A script running in parallel uses iptables to interrupt (DROP or REJECT chosen randomly) at random times for a few seconds the LDAP connection.

After running the above for some time, verify that readlog and writelog are consistent, i.e. that each value in readlog corresponds to the writelog entry that most closely precedes the readlog entry (give or take a few milliseconds in case several writes were issued in parallel at almost the same time).

**Summary:** Load test/stress test for jobdb and possibly LDAP

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on February 18, 2013 10:04:37_

=> Bonus

**Labels:** -Priority-Normal Priority-Bonus

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on February 18, 2013 10:57:49_

Status => Accepted

**Status:** Accepted

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on March 22, 2013 02:09:12_

Possibly relevant for MS 5

**Labels:** Milestone-5

**Comment by mbenkmann** [comment]
_From [matthias...@gmail.com](https://code.google.com/u/110029356478037175549/) on July 06, 2013 05:32:29_

A good idea, but there will always be something more important to push this back. As an artificial test there is simply no compelling necessity for this. => WONTFIX

**Status:** WontFix
